### PR TITLE
Register common variables rather than passing them in all the time

### LINKED ISSUE #137: Refactor callWithTelemetryAndErrorHandling so that you only have to pass in callbackId and callback
We should do some sort of 'preregistration' of the telemetryReporter, outputChannel, and extensionContext rather than forcing people to pass them in every time

### PR DESCRIPTION
**Pull request by EricJizbaMSFT**
Before I did the work for the user-agent (which will require extensionContext to be passed a bunch of places to get the package version), I decided to fix #137. This allows us to use common variables without having to pass them in everywhere.

I implemented this all the way through for the functions extension to make sure it would work with tests, etc. Assuming you approve, I'll start fixing in the other extensions as well.

### CODE REVIEW

**Comment by nturinski** [comment]
This is replacing extensionVariables in some of our extensions then?  I'm pro this.

**Comment by EricJizbaMSFT** [comment]
Okay I've sent out all PRs to give a full example using the functions extension: https://github.com/Microsoft/vscode-azurefunctions/pull/434

I would say this _leverages_ the extensionVariables, rather than _replaces_ the extensionVariables. I wanted to keep the ability for each extension/package to have different variables that they use. That's why 'ext' is not exposed from the ui package. You can register the exact same 'ext', but the ui package will only use the 4 variables that it needs from within the package.

File: ui/index.d.ts [hunk 1]
```diff
@@ -620,3 +615,18 @@ export declare namespace DialogResponses {
...
+ */
+export declare function registerUIExtensionVariables(extVars: UIExtensionVariables): void;
+
+/**
+ * Interface for common extension variables used throughout the UI package.
+ */
+export interface UIExtensionVariables {
+    context: ExtensionContext;
+    outputChannel: OutputChannel;
+    ui: IAzureUserInput;
```
Comment (StephenWeatherford) [inline ui/index.d.ts]: Tests will need to change the IAzureUserInput, right?  Or do they refill it with data?

File: ui/src/AzureActionHandler.ts [hunk 2]
```diff
@@ -3,54 +3,37 @@
  *  Licensed under the MIT License. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
-import { commands, Event, ExtensionContext, OutputChannel } from 'vscode';
-import TelemetryReporter from 'vscode-extension-telemetry';
+import { commands, Event } from 'vscode';
 import { IActionContext } from '../index';
 import { callWithTelemetryAndErrorHandling } from './callWithTelemetryAndErrorHandling';
+import { ext } from './extensionVariables';
```
Comment (StephenWeatherford) [inline ui/src/AzureActionHandler.ts]: This does seem a bit confusing with extension vars, is there a better name?

File: ui/index.d.ts (hunk 1 above)
Comment (StephenWeatherford) [reply ui/index.d.ts]: Oh, you're not making a copy, you're keeping a reference...

File: ui/src/AzureActionHandler.ts (hunk 2 above)
Comment (StephenWeatherford) [reply ui/src/AzureActionHandler.ts]: Never mind, makes more sense now.

File: ui/index.d.ts (hunk 1 above)
Comment (EricJizbaMSFT) [reply ui/index.d.ts]: Yep it's easy for tests to change it since it's a reference

**Pull request merged**